File: leetcode-cn/Python/77.组合.py
#
# @lc app=leetcode.cn id=77 lang=python3
#
# [77] 组合
#

# @lc code=start
from typing import List


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        def help(start, k):
            if k == 0 and len(path) == m:
                result.append(path[:])
            for i in range(start, n + 1):
                path.append(i)
                help(i + 1, k - 1)
                path.pop()

        path, result = [], []
        m = k
        help(1, k)

        return result


# @lc code=end

File: leetcode-cn/C++/77.组合.cpp
/*
 * @lc app=leetcode.cn id=77 lang=cpp
 *
 * [77] 组合
 */

// @lc code=start
#include <vector>
using namespace std;

class Solution
{
public:
    vector<vector<int>> ans;
    vector<int> path;
    int k = 0;
    int n = 0;

    void help(int index)
    {
        if (path.size() == k)
        {
            ans.push_back(path);
            return;
        }

        for (int i = index; i <= n; i++)
        {
            path.push_back(i);
            help(i + 1);
            path.pop_back();
        }
    }

    vector<vector<int>> combine(int n, int k)
    {
        this->k = k;
        this->n = n;
        help(1);
        return ans;
    }
};
// @lc code=end
